packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -0,0 +1,107 @@
+ * but the actual value specified is irrelevant
+ */
+export const REQUIRED_BY_CFN = 'required-to-be-present-by-cfn';
+
+export async function isHotswappableS3BucketDeploymentChange(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapResult> {
+  // In old-style synthesis, the policy used by the lambda to copy assets Ref's the assets directly,
+  // meaning that the changes made to the Policy are artifacts that can be safely ignored
+  if (change.newValue.Type === 'AWS::IAM::Policy') {
+    const roles = change.newValue.Properties?.Roles;
+    for (const role of roles) {
+      const roleLogicalName = await evaluateCfnTemplate.findLogicallNameFor(await evaluateCfnTemplate.evaluateCfnExpression(role));
+      if (!roleLogicalName) {
+        return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+      }
+
+      const roleRefs = evaluateCfnTemplate.findReferencesTo(roleLogicalName);
+      for (const roleRef of roleRefs) {
+        if (roleRef.Type === 'AWS::Lambda::Function') {
+          const lambdaRefs = evaluateCfnTemplate.findReferencesTo(roleRef.LogicalId);
+          for (const lambdaRef of lambdaRefs) {
+            // If S3Deployment -> Lambda -> Role and IAM::Policy -> Role, then this IAM::Policy change is an
+            // artifact of old-style synthesis
+            if (lambdaRef.Type === 'Custom::CDKBucketDeployment') {
+              return new EmptyHotswapOperation();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  if (change.newValue.Type !== 'Custom::CDKBucketDeployment') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  // note that this gives the ARN of the lambda, not the name. This is fine though, the invoke() sdk call will take either
+  const functionName = await establishResourcePhysicalName(logicalId, change.newValue.Properties?.ServiceToken, evaluateCfnTemplate);
+  if (!functionName) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const sourceBucketNames = await evaluateCfnTemplate.evaluateCfnExpression(change.newValue.Properties?.SourceBucketNames);
+  const sourceObjectKeys = await evaluateCfnTemplate.evaluateCfnExpression(change.newValue.Properties?.SourceObjectKeys);
+  const destinationBucketName = await evaluateCfnTemplate.evaluateCfnExpression(change.newValue.Properties?.DestinationBucketName);
+  const destinationBucketKeyPrefix = await evaluateCfnTemplate.evaluateCfnExpression(change.newValue.Properties?.DestinationBucketKeyPrefix);
+
+  return new S3BucketDeploymentHotswapOperation({
+    FunctionName: functionName,
+    SourceBucketNames: sourceBucketNames,
+    SourceObjectKeys: sourceObjectKeys,
+    DestinationBucketName: destinationBucketName,
+    DestinationBucketKeyPrefix: destinationBucketKeyPrefix,
+  });
+}
+
+interface S3BucketDeployment {
+  FunctionName: string,
```

Comment:
```suggestion
  FunctionName: string;
```

Same comment on lines below.